File: src/main/java/com/milaboratory/mixcr/cli/CommandExportOverlap.java
package com.milaboratory.mixcr.cli;

import cc.redberry.pipe.CUtils;
import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.basictypes.CloneReader;
import com.milaboratory.mixcr.basictypes.CloneSetIO;
import com.milaboratory.mixcr.export.FieldExtractor;
import com.milaboratory.mixcr.export.OutputMode;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapDataset;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapGroup;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapUtil;
import com.milaboratory.mixcr.util.OutputPortWithProgress;
import com.milaboratory.util.CanReportProgress;
import com.milaboratory.util.SmartProgressReporter;
import io.repseq.core.Chains;
import io.repseq.core.Chains.NamedChains;
import io.repseq.core.GeneFeature;
import io.repseq.core.GeneType;
import io.repseq.core.VDJCLibraryRegistry;
import picocli.CommandLine.Command;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Command(separator = " ",
        description = "Build cloneset overlap and export into tab delimited file."
)
public class CommandExportOverlap extends ACommandWithOutputMiXCR {
    @Parameters(description = "cloneset.{clns|clna}...")
    public List<String> inOut;

    @Option(description = "Chains to export",
            names = "--chains")
    public List<String> chains = null;

    @Option(description = "Filter out-of-frame sequences and clonotypes with stop-codons",
            names = {"--only-productive"})
    public boolean onlyProductive = false;

    @Option(description = "Overlap criteria. Default CDR3|AA|V|J",
            names = {"--criteria"})
    public String overlapCriteria = "CDR3|AA|V|J";

    /** auto-generated opts (exporters) injected manually */
    private CommandSpec spec;

    @Override
    public List<String> getInputFiles() {
        return inOut.subList(0, inOut.size() - 1);
    }

    public Path getOut(NamedChains chains) {
        Path out = Paths.get(inOut.get(inOut.size() - 1)).toAbsolutePath();
        if (chains == Chains.ALL_NAMED)
            return out;
        String fName = out.getFileName().toString();
        if (fName.endsWith(".tsv"))
            fName = fName.replace("tsv", "") + chains.name + ".tsv";
        else
            fName = fName + "_" + chains.name;
        return out.getParent().resolve(fName);
    }

    @Override
    public void run0() throws Exception {
        List<String> samples = getInputFiles();
        boolean needRecalculateCount = this.chains != null || onlyProductive;

        List<NamedChains> chains = this.chains == null
                ? Collections.singletonList(Chains.ALL_NAMED)
                : this.chains.stream().map(Chains::getNamedChains).collect(Collectors.toList());
        Map<NamedChains, double[]> counts = null;
        if (needRecalculateCount) {
            counts = new HashMap<>();
            AtomicInteger sampleIndex = new AtomicInteger(0);
            SmartProgressReporter.startProgressReport("Calculating samples counts", new CanReportProgress() {
                @Override
                public double getProgress() {
                    return 1.0 * sampleIndex.get() / samples.size();
                }

                @Override
                public boolean isFinished() {
                    return sampleIndex.get() == samples.size();
                }
            });
            for (int i = 0; i < samples.size(); i++) {
                try (CloneReader reader = CloneSetIO.mkReader(Paths.get(samples.get(i)), VDJCLibraryRegistry.getDefault())) {
                    for (Clone cl : CUtils.it(reader.readClones())) {
                        if (!isProductive(cl))
                            continue;
                        for (NamedChains ch : chains) {
                            if (!hasChains(cl, ch))
                                continue;
                            counts.computeIfAbsent(ch, __ -> new double[samples.size()])[i] += cl.getCount();
                        }
                    }
                }
                sampleIndex.set(i + 1);
            }
        }

//        try (OutputPortWithProgress<OverlapGroup<Clone>> port = overlap.mkElementsPort()) {
//            SmartProgressReporter.startProgressReport("Computing total counts", port);
//            for (OverlapGroup<Clone> row : CUtils.it(port)) {
//                filterProductive(row, true);
//                for (NamedChains ch : chains) {
//                    double[] c = counts.computeIfAbsent(ch.chains, __ -> new double[getInputFiles().size()]);
//                    OverlapGroup<Clone> projected = forChains(row, ch.chains);
//                    for (int i = 0; i < projected.size(); i++) {
//                        c[i] += projected.getBySample(i).stream().mapToDouble(Clone::getCount).sum();
//                    }
//                }
//            }
//        }
        OverlapUtil.OverlapCriteria criteria = OverlapUtil.parseCriteria(this.overlapCriteria);
        List<OverlapFieldExtractor> extractors = new ArrayList<>();

        extractors.add(new ExtractorUnique(
                new FieldExtractor<Clone>() {
                    @Override
                    public String getHeader() {
                        return (criteria.isAA ? "aaSeq" : "nSeq") + GeneFeature.encode(criteria.feature);
                    }

                    @Override
                    public String extractValue(Clone object) {
                        return criteria.isAA
                                ? object.getAAFeature(criteria.feature).toString()
                                : object.getNFeature(criteria.feature).toString();
                    }
                }
        ));
        if (criteria.withV)
            extractors.add(new ExtractorUnique(
                    new FieldExtractor<Clone>() {
                        @Override
                        public String getHeader() {
                            return "vGene";
                        }

                        @Override
                        public String extractValue(Clone object) {
                            return object.getBestHit(GeneType.Variable).getGene().getName();
                        }
                    }
            ));
        if (criteria.withJ)
            extractors.add(new ExtractorUnique(
                    new FieldExtractor<Clone>() {
                        @Override
                        public String getHeader() {
                            return "jGene";
                        }

                        @Override
                        public String extractValue(Clone object) {
                            return object.getBestHit(GeneType.Joining).getGene().getName();
                        }
                    }
            ));

        extractors.add(new NumberOfSamples());
        extractors.add(new TotalCount());
        extractors.add(new TotalFraction());

        List<FieldExtractor<Clone>> fieldExtractors = CommandExport
                .parseSpec(spec.commandLine().getParseResult())
                .stream()
                .flatMap(f -> CommandExport.extractor(f, Clone.class, OutputMode.ScriptingFriendly).stream())
                .collect(Collectors.toList());

        for (FieldExtractor<? super Clone> fe : fieldExtractors)
            extractors.add(new ExtractorPerSample(fe));

        Map<NamedChains, InfoWriter> writers = new HashMap<>();
        for (NamedChains chain : chains) {
            InfoWriter writer = new InfoWriter(getOut(chain), samples, extractors);
            writer.writeHeader();
            writers.put(chain, writer);
        }
        OverlapDataset<Clone> overlap = OverlapUtil.overlap(samples, criteria.ordering());
        try (OutputPortWithProgress<OverlapGroup<Clone>> port = overlap.mkElementsPort()) {
            SmartProgressReporter.startProgressReport("Calculating overlap", port);
            for (OverlapGroup<Clone> row : CUtils.it(port)) {
                row = filterProductive(row, true);
                if (row == null)
                    continue;

                for (NamedChains ch : chains) {
                    OverlapGroup<Clone> forChain = forChains(row, ch);
                    if (forChain == null)
                        continue;

                    if (needRecalculateCount) {
                        double[] cs = counts.get(ch);
                        for (int i = 0; i < forChain.size(); i++)
                            for (Clone cl : forChain.getBySample(i))
                                cl.overrideFraction(cl.getCount() / cs[i]);
                    }

                    writers.get(ch).writeRow(forChain);
                }
            }
        }
        writers.forEach((___, w) -> w.close());
    }

    private static OverlapGroup<Clone> filter(OverlapGroup<Clone> row, Predicate<Clone> criteria, boolean inPlace) {
        if (inPlace) {
            boolean empty = true;
            for (List<Clone> l : row) {
                l.removeIf(c -> !criteria.test(c));
                if (!l.isEmpty())
                    empty = false;
            }
            if (empty)
                return null;
            else
                return row;
        } else {
            boolean empty = true;
            List<List<Clone>> r = new ArrayList<>();
            for (List<Clone> l : row) {
                List<Clone> f = l.stream().filter(criteria).collect(Collectors.toList());
                r.add(f);
                if (!f.isEmpty())
                    empty = false;
            }
            if (empty)
                return null;
            else
                return new OverlapGroup<>(r);
        }
    }

    private static boolean isProductive(Clone c) {
        return !c.isOutOfFrame(GeneFeature.CDR3) && !c.containsStops(GeneFeature.CDR3);
    }

    private static boolean hasChains(Clone c, NamedChains chains) {
        return chains == Chains.ALL_NAMED || Arrays.stream(GeneType.VJC_REFERENCE).anyMatch(gt -> {
            Chains clChains = c.getAllChains(gt);
            if (clChains == null)
                return false;
            return clChains.intersects(chains.chains);
        });
    }

    private static OverlapGroup<Clone> forChains(OverlapGroup<Clone> row, NamedChains chains) {
        return chains == Chains.ALL_NAMED ? row : filter(row, c -> hasChains(c, chains), false);
    }

    private OverlapGroup<Clone> filterProductive(OverlapGroup<Clone> row, boolean inPlace) {
        return onlyProductive ? filter(row, CommandExportOverlap::isProductive, inPlace) : row;
    }

    private static final class InfoWriter implements AutoCloseable {
        final PrintWriter writer;
        final List<String> samples;
        final List<OverlapFieldExtractor> extractors;

        public InfoWriter(Path out, List<String> samples,
                          List<OverlapFieldExtractor> extractors) {
            try {
                this.writer = new PrintWriter(new FileWriter(out.toFile()));
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
            this.samples = samples;
            this.extractors = extractors;
        }

        void writeHeader() {
            writer.println(extractors.stream().flatMap(e -> e.header(samples).stream()).collect(Collectors.joining("\t")));
        }

        void writeRow(OverlapGroup<Clone> row) {
            writer.println(extractors.stream().flatMap(e -> e.values(row).stream()).collect(Collectors.joining("\t")));
        }

        @Override
        public void close() {
            writer.close();
        }
    }

    private interface OverlapFieldExtractor {
        List<String> header(List<String> samples);

        List<String> values(OverlapGroup<Clone> row);
    }

    private static final class TotalCount implements OverlapFieldExtractor {
        @Override
        public List<String> header(List<String> samples) {
            return samples.stream().map(s -> s + "_countAggregated").collect(Collectors.toList());
        }

        @Override
        public List<String> values(OverlapGroup<Clone> row) {
            return row.stream().map(l -> Double.toString(l.stream().mapToDouble(Clone::getCount).sum()))
                    .collect(Collectors.toList());
        }
    }

    private static final class TotalFraction implements OverlapFieldExtractor {
        @Override
        public List<String> header(List<String> samples) {
            return samples.stream().map(s -> s + "_fractionAggregated").collect(Collectors.toList());
        }

        @Override
        public List<String> values(OverlapGroup<Clone> row) {
            return row.stream().map(l -> Double.toString(l.stream().mapToDouble(Clone::getFraction).sum()))
                    .collect(Collectors.toList());
        }
    }

    private static final class NumberOfSamples implements OverlapFieldExtractor {
        @Override
        public List<String> header(List<String> samples) {
            return Collections.singletonList("nSamples");
        }

        @Override
        public List<String> values(OverlapGroup<Clone> row) {
            return Collections.singletonList("" + row.stream().filter(l -> l.size() > 0).count());
        }
    }

    private static final class ExtractorUnique implements OverlapFieldExtractor {
        final FieldExtractor<Clone> extractor;

        public ExtractorUnique(FieldExtractor<Clone> extractor) {
            this.extractor = extractor;
        }

        @Override
        public List<String> header(List<String> samples) {
            return Collections.singletonList(extractor.getHeader());
        }

        @Override
        public List<String> values(OverlapGroup<Clone> row) {
            return Collections.singletonList(extractor.extractValue(row.stream().filter(l -> l.size() > 0).findAny().get().get(0)));
        }
    }

    private static final class ExtractorPerSample implements OverlapFieldExtractor {
        final FieldExtractor<? super Clone> extractor;

        public ExtractorPerSample(FieldExtractor<? super Clone> extractor) {
            this.extractor = extractor;
        }

        @Override
        public List<String> header(List<String> samples) {
            return samples.stream().map(s -> s + "_" + extractor.getHeader()).collect(Collectors.toList());
        }

        @Override
        public List<String> values(OverlapGroup<Clone> row) {
            return row.stream().map(s ->
                    s.stream().map(extractor::extractValue).collect(Collectors.joining(","))
            ).collect(Collectors.toList());
        }
    }

    public static CommandSpec mkSpec() {
        CommandExportOverlap export = new CommandExportOverlap();
        CommandSpec spec = CommandSpec.forAnnotatedObject(export);
        export.spec = spec; // inject spec manually
        CommandExport.addOptionsToSpec(spec, Clone.class);
        return spec;
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandExportOverlap.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.basictypes.IOUtil
import com.milaboratory.mixcr.basictypes.tag.TagType
import com.milaboratory.mixcr.cli.CommonDescriptions.Labels
import com.milaboratory.mixcr.export.CloneFieldsExtractorsFactory
import com.milaboratory.mixcr.export.ExportFieldDescription
import com.milaboratory.mixcr.export.FieldExtractor
import com.milaboratory.mixcr.export.HeaderForExport
import com.milaboratory.mixcr.export.RowMetaForExport
import com.milaboratory.mixcr.postanalysis.overlap.OverlapGroup
import com.milaboratory.mixcr.postanalysis.overlap.OverlapUtil
import com.milaboratory.mixcr.postanalysis.preproc.ChainsFilter
import com.milaboratory.mixcr.postanalysis.util.OverlapBrowser
import com.milaboratory.primitivio.forEach
import com.milaboratory.util.SmartProgressReporter
import io.repseq.core.Chains
import io.repseq.core.GeneFeature
import io.repseq.core.GeneType.Joining
import io.repseq.core.GeneType.Variable
import picocli.CommandLine.Command
import picocli.CommandLine.Help.Visibility.ALWAYS
import picocli.CommandLine.Model
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.io.FileWriter
import java.io.PrintWriter
import java.nio.file.Path
import kotlin.collections.component1
import kotlin.collections.component2

@Command(
    description = ["Build cloneset overlap and export into tab delimited file."],
)
class CommandExportOverlap : MiXCRCommandWithOutputs() {
    companion object {
        private const val inputsLabel = "cloneset.(clns|clna)..."

        private const val outputLabel = "output.tsv"

        @JvmStatic
        fun mkSpec(): Model.CommandSpec {
            val export = CommandExportOverlap()
            val spec = Model.CommandSpec.forAnnotatedObject(export)
                .addPositional(
                    Model.PositionalParamSpec.builder()
                        .index("0")
                        .required(false)
                        .arity("0..*")
                        .type(Path::class.java)
                        .paramLabel(inputsLabel)
                        .hideParamSyntax(true)
                        .description("Paths to input files")
                        .build()
                )
                .addPositional(
                    Model.PositionalParamSpec.builder()
                        .index("1")
                        .required(false)
                        .arity("0..*")
                        .type(Path::class.java)
                        .paramLabel(outputLabel)
                        .hideParamSyntax(true)
                        .description(
                            "Path template where to write output export tables.",
                            "For each `chain` will be generated table with path `{outputDir}/{outputFileName}.{chain}.tsv`"
                        )
                        .build()
                )

            export.spec = spec // inject spec manually
            CloneFieldsExtractorsFactory.addOptionsToSpec(export.addedFields, spec)
            return spec
        }
    }

    @Parameters(
        index = "0",
        arity = "2..*",
        paramLabel = "$inputsLabel $outputLabel",
        hideParamSyntax = true,
        //help is covered by mkCommandSpec
        hidden = true
    )
    var inOut: List<Path> = mutableListOf()

    @Option(
        description = ["Output overlap for specified chains only; if multiple chains are specified, results per each chains will be exported in separate files."],
        names = ["--chains"],
        split = ",",
        paramLabel = Labels.CHAIN
    )
    var chains: Set<String>? = null

    @Option(
        description = [CommonDescriptions.OVERLAP_CRITERIA],
        names = ["--criteria"],
        showDefaultValue = ALWAYS,
        paramLabel = "<s>"
    )
    var overlapCriteria = "CDR3|AA|V|J"

    @Option(
        description = ["Filter out-of-frame sequences and clonotypes with stop-codons"],
        names = ["--only-productive"]
    )
    var onlyProductive = false

    private val output get() = inOut.last()

    public override val inputFiles
        get() = inOut.subList(0, inOut.size - 1)

    override val outputFiles
        get() = listOf(output)

    private fun getOut(chains: Chains): Path {
        val fName = output.fileName.toString().replace("tsv", "${chains}.tsv")
        return output.toAbsolutePath().parent.resolve(fName)
    }

    override fun validate() {
        inputFiles.forEach { input ->
            ValidationException.requireFileType(input, InputFileType.CLNX)
        }
        ValidationException.requireFileType(output, InputFileType.TSV)
    }

    private val tagsInfo
        get() = inputFiles.map { IOUtil.extractHeader(it).tagsInfo }

    private fun tagLevel(tt: TagType) =
        if (tagsInfo.all { it.hasTagsWithType(tt) }) {
            val set = tagsInfo.map { it.getDepthFor(tt) }.toSet()
            if (set.size > 1)
                null
            else
                set.first()
        } else
            null

    var addedFields: MutableList<ExportFieldDescription> = mutableListOf()

    override fun run0() {
        val samples = inputFiles
        val chains = this.chains?.let { ChainsFilter.parseChainsList(this.chains) }
        val criteria = OverlapUtil.parseCriteria(overlapCriteria)
        val extractors = mutableListOf<OverlapFieldExtractor>()
        extractors += ExtractorUnique(
            object : FieldExtractor<Clone> {
                override val header = (if (criteria.isAA) "aaSeq" else "nSeq") + GeneFeature.encode(criteria.feature)

                override fun extractValue(header: RowMetaForExport, obj: Clone) = when {
                    criteria.isAA -> obj.getAAFeature(criteria.feature).toString()
                    else -> obj.getNFeature(criteria.feature).toString()
                }
            }
        )
        if (criteria.withV) {
            extractors += ExtractorUnique(
                object : FieldExtractor<Clone> {
                    override val header = "vGene"

                    override fun extractValue(header: RowMetaForExport, obj: Clone) = obj.getBestHit(Variable).gene.name
                }
            )
        }
        if (criteria.withJ) {
            extractors += ExtractorUnique(
                object : FieldExtractor<Clone> {
                    override val header: String = "jGene"

                    override fun extractValue(header: RowMetaForExport, obj: Clone) = obj.getBestHit(Joining).gene.name
                }
            )
        }
        extractors += NumberOfSamples()
        extractors += TotalCount()
        extractors += TotalFraction()

        tagLevel(TagType.Molecule)?.let { umiLevel ->
            extractors += TotalTagCount(umiLevel, "UMI")
            extractors += TotalTagFraction(umiLevel, "UMI")
        }

        tagLevel(TagType.Cell)?.let { cellLevel ->
            extractors += TotalTagCount(cellLevel, "CELL")
            extractors += TotalTagFraction(cellLevel, "CELL")
        }

        val header = IOUtil.extractHeader(samples[0])
        val fieldExtractors: List<FieldExtractor<Clone>> = CloneFieldsExtractorsFactory.createExtractors(
            addedFields,
            HeaderForExport(listOf(header.tagsInfo), header.allFullyCoveredBy)
        )

        extractors += fieldExtractors.map { ExtractorPerSample(it) }

        val overlapBrowser = OverlapBrowser(onlyProductive)
        SmartProgressReporter.startProgressReport(overlapBrowser)

        val countsByChain = overlapBrowser.computeCountsByChain(samples)
        val chainsToWrite = chains?.intersect(countsByChain.keys) ?: countsByChain.keys
        val writers = chainsToWrite.associateWith { chain ->
            val writer = InfoWriter(getOut(chain), samples, extractors)
            writer.writeHeader()
            writer
        }

        val overlap = OverlapUtil.overlap(samples.map { it.toString() }, { true }, criteria.ordering())
        val rowMetaForExport = RowMetaForExport(header.tagsInfo)
        overlap.mkElementsPort().use { port ->
            overlapBrowser.overlap(countsByChain, port).forEach { row ->
                for ((ch, cloneOverlapGroup) in row) {
                    writers[ch]!!.writeRow(rowMetaForExport, cloneOverlapGroup)
                }
            }
        }
        writers.values.forEach { it.close() }
    }

    private class InfoWriter(
        out: Path,
        samples: List<Path>,
        private val extractors: List<OverlapFieldExtractor>
    ) : AutoCloseable {
        private val writer: PrintWriter = PrintWriter(FileWriter(out.toFile()))
        private val samplesNames = samples.map { sampleName -> removeExt(sampleName.toString()) }

        fun writeHeader() {
            writer.println(extractors
                .flatMap { fieldExtractor -> fieldExtractor.header(samplesNames) }
                .joinToString("\t")
            )
        }

        fun writeRow(header: RowMetaForExport, row: OverlapGroup<Clone>) {
            writer.println(extractors
                .flatMap { fieldExtractor -> fieldExtractor.values(header, row) }
                .joinToString("\t")
            )
        }

        override fun close() {
            writer.close()
        }

        companion object {
            private fun removeExt(sampleName: String): String = sampleName
                .replace(".clns", "")
                .replace(".clna", "")
        }
    }

    private interface OverlapFieldExtractor {
        fun header(samples: List<String>): List<String>
        fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String>
    }

    private class TotalCount : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> =
            samples.map { sample -> "${sample}_readCountAggregated" }

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            row.map { clones -> clones.sumOf { it.count }.toString() }
    }

    private class TotalFraction : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> =
            samples.map { sample -> "${sample}_readFractionAggregated" }

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            row.map { clones -> clones.sumOf { it.fraction }.toString() }
    }

    private class TotalTagCount(val level: Int, val tagName: String) : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> =
            samples.map { sample -> "${sample}_unique${tagName}CountAggregated" }

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            row.map { clones -> clones.sumOf { it.getTagDiversity(level) }.toString() }
    }

    private class TotalTagFraction(val level: Int, val tagName: String) : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> =
            samples.map { sample -> "${sample}_unique${tagName}FractionAggregated" }

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            row.map { clones -> clones.sumOf { it.getTagDiversityFraction(level) }.toString() }
    }

    private class NumberOfSamples : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> = listOf("nSamples")

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            listOf(row.count { clones -> clones.isNotEmpty() }.toString())
    }

    private class ExtractorUnique(
        private val extractor: FieldExtractor<Clone>
    ) : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> = listOf(extractor.header)

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            listOf(extractor.extractValue(header, row.first { clones -> clones.isNotEmpty() }.first()))
    }

    private class ExtractorPerSample(val extractor: FieldExtractor<Clone>) : OverlapFieldExtractor {
        override fun header(samples: List<String>): List<String> =
            samples.map { sample -> sample + "_" + extractor.header }

        override fun values(header: RowMetaForExport, row: OverlapGroup<Clone>): List<String> =
            row.map { clones ->
                clones.joinToString(",") { clone -> extractor.extractValue(header, clone) }
            }
    }
}
